File: moq-web/src/internal/message/track_info_test.ts
import { assertEquals } from "@std/assert";
import { TrackInfoMessage } from "./track_info.ts";
import { Buffer } from "@okdaichi/golikejs/bytes";

Deno.test("TrackInfoMessage - encode/decode roundtrip", async (t) => {
	const testCases = {
		"millisecond timescale": {
			publisherPriority: 5,
			publisherOrdered: 1,
			publisherMaxLatency: 2000,
			timescale: 1000,
		},
		"rtp video clock": {
			publisherPriority: 128,
			publisherOrdered: 0,
			publisherMaxLatency: 0,
			timescale: 90000,
		},
	};

	for (const [caseName, input] of Object.entries(testCases)) {
		await t.step(caseName, async () => {
			const buffer = Buffer.make(64);
			const message = new TrackInfoMessage(input);
			assertEquals(await message.encode(buffer), undefined);

			const decoded = new TrackInfoMessage({});
			assertEquals(await decoded.decode(buffer), undefined);
			assertEquals(decoded.publisherPriority, input.publisherPriority);
			assertEquals(decoded.publisherOrdered, input.publisherOrdered);
			assertEquals(decoded.publisherMaxLatency, input.publisherMaxLatency);
			assertEquals(decoded.timescale, input.timescale);
		});
	}

	await t.step("decode returns error on empty input", async () => {
		const buffer = Buffer.make(0);
		const decoded = new TrackInfoMessage({});
		const err = await decoded.decode(buffer);
		assertEquals(err !== undefined, true);
	});
});

File: moqt/internal/message/track_info_test.go
package message_test

import (
	"bytes"
	"testing"

	"github.com/qumo-dev/gomoqt/moqt/internal/message"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestTrackInfoMessage_EncodeDecode(t *testing.T) {
	tests := map[string]struct {
		input message.TrackInfoMessage
	}{
		"millisecond timescale": {
			input: message.TrackInfoMessage{
				PublisherPriority:   5,
				PublisherOrdered:    1,
				PublisherMaxLatency: 2000,
				Timescale:           1000,
			},
		},
		"rtp video clock": {
			input: message.TrackInfoMessage{
				PublisherPriority: 128,
				Timescale:         90000,
			},
		},
	}

	for name, tt := range tests {
		t.Run(name, func(t *testing.T) {
			var buf bytes.Buffer

			require.NoError(t, tt.input.Encode(&buf))

			var decoded message.TrackInfoMessage
			require.NoError(t, decoded.Decode(&buf))

			assert.Equal(t, tt.input, decoded)
		})
	}
}

func TestTrackInfoMessage_DecodeErrors(t *testing.T) {
	tests := map[string]struct {
		data []byte
	}{
		"empty input":     {data: []byte{}},
		"too short":       {data: []byte{0x01, 0x05}},
		"trailing bytes": {data: []byte{0x05, 0x01, 0x00, 0x00, 0x00, 0xFF}},
	}

	for name, tt := range tests {
		t.Run(name, func(t *testing.T) {
			var decoded message.TrackInfoMessage
			err := decoded.Decode(bytes.NewReader(tt.data))
			assert.Error(t, err)
		})
	}
}
